src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala

```diff
@@ -29,154 +29,121 @@ class TaskReplaceActor(
+    // reset the launch queue delay
+    logger.info("Resetting the backoff delay before restarting the runSpec")
+    launchQueue.resetDelay(runSpec)
 
-  // Only old instances that still have the Goal.Running
-  val oldActiveInstances = oldInstances.filter(_.state.goal == Goal.Running)
+    // Fetch state of apps
+    val system = akka.actor.ActorSystem("system")
+    tick = system.scheduler.schedule(0 seconds, 10 seconds)(request_health_status)
+  }
 
-  // All instances to kill as set for quick lookup
-  private[this] var oldActiveInstanceIds: SortedSet[Id] = oldActiveInstances.map(_.instanceId).to[SortedSet]
+  def isHealthy(instance: Instance, healths: Map[Instance.Id, Seq[Health]]): Boolean = {
+    // FIXME: Or use state.healthy.getOrElse
+    val i_health = healths.find(_._1 == instance.instanceId)
+    if (instance.hasConfiguredHealthChecks && i_health.isDefined) { // Has healthchecks
+      if (_isHealthy(i_health.get._2) == true) {
+        logger.info(s"Checking ${instance.instanceId}: has healcheck and is healthy")
+        return true
+      } else {
+        logger.info(s"Checking ${instance.instanceId}: has healcheck and is NOT healthy")
+        return false
+      }
+    } else if (instance.hasConfiguredHealthChecks && !i_health.isDefined) {
+      logger.info(s"Checking ${instance.instanceId}: has healcheck and is UNKNOWN")
+      return false
+    } else { // Don't have healthchecks
+      logger.info(s"Checking ${instance.instanceId}: has now HC")
+      if (instance.state.goal == Goal.Running) return true
+      else return false
+    }
+  }
 
-  // All instances to kill queued up
-  private[this] val toKill: mutable.Queue[Instance.Id] = oldActiveInstances.map(_.instanceId).to[mutable.Queue]
+  def _isHealthy(healths: Seq[Health]): Boolean = {
+    for (h <- healths)
+      if (!h.alive) return false
+    return true
+  }
 
-  // The number of started instances. Defaults to the number of already started instances.
-  var instancesStarted: Int = instancesAlreadyStarted.size
+  def request_health_status(): Unit = {
+    logger.info(s"Requesting HealthStatus for ${pathId}")
+    deploymentManagerActor ! HealthStatusRequest(pathId)
+  }
 
-  @SuppressWarnings(Array("all")) // async/await
-  override def preStart(): Unit = {
-    super.preStart()
-    // subscribe to all needed events
-    eventBus.subscribe(self, classOf[InstanceChanged])
-    eventBus.subscribe(self, classOf[InstanceHealthChanged])
+  def step(health: Map[Instance.Id, Seq[Health]]): Unit = {
+    logger.info(s"---=== DEPLOYMENT STEP FOR ${pathId} ===---yy")
+    val current_instances = instanceTracker.specInstancesSync(pathId)
+    var new_running = 0
+    var old_running = 0
+    var new_failing = 0
+    var old_failing = 0
+    for (i <- current_instances) {
+      if (i.runSpecVersion == runSpec.version) { // New version
+        if (isHealthy(i, health)) new_running += 1
```

Comment:
Question: do we know how "readyness check" play into all of this?

Reply:
I admit it's a bit blurry for now. This is one of the thing I still need to test. I need to dig the health status behavior to see if I can get a state of readiness behavior.
(I won't merge this PR until this is clear)

Reply:
as far as I know, we don't use readyness check internally but this will be a question for upstream.

Reply:
This is covered in the 3rd (and last) test suite I'm fixing :)